enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/ImageProcessor.java

```diff
@@ -0,0 +1,106 @@
+
+import com.wultra.app.enrollmentserver.model.integration.Image;
+import com.wultra.app.enrollmentserver.model.integration.OwnerId;
+import com.wultra.app.onboardingserver.errorhandling.PresenceCheckException;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+import javax.imageio.ImageIO;
+import javax.imageio.ImageReader;
+import javax.imageio.stream.ImageInputStream;
+import java.awt.image.BufferedImage;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+
+/**
+ * Component to image processing.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+@Component
+@Slf4j
+public class ImageProcessor {
+
+    /**
+     * Upscale the given image if its width is smaller than the required value, otherwise the original image is returned.
+     *
+     * @param ownerId Owner identification for logging purpose
+     * @param sourceImage the image to upscale
+     * @param minimalWidth required minimal width of the image in pixels
+     * @return the upscaled image if needed or the original
+     */
+    public Image upscaleImage(final OwnerId ownerId, final Image sourceImage, final int minimalWidth) throws PresenceCheckException {
+        final String filename = sourceImage.getFilename();
+        logger.debug("Attempt to upscale image: {} to minimalWidth: {} px, {}", filename, minimalWidth, ownerId);
+
+        try {
+            final BufferedImageWrapper bufferedSourceImageWrapper = readImage(new ByteArrayInputStream(sourceImage.getData()));
+            final BufferedImage bufferedSourceImage = bufferedSourceImageWrapper.bufferedImage;
+            final String formatName = bufferedSourceImageWrapper.formatName;
+            final int currentWidth = bufferedSourceImage.getWidth();
+            final int currentHeight = bufferedSourceImage.getHeight();
+            logger.info("Processing image: {}, format: {}, size: {} x {} px, {}", filename, formatName, currentWidth, currentHeight, ownerId);
+
+            if (currentWidth < minimalWidth) {
+                final double aspectRation = minimalWidth / (double) currentWidth;
+                final int minimalHeight = (int) (aspectRation * currentHeight);
+                logger.info("Upscaling image: {} to minimal size: {} x {} px, {}", filename, minimalWidth, minimalHeight, ownerId);
+                final java.awt.Image upscaledImage = bufferedSourceImage.getScaledInstance(minimalWidth, minimalHeight, java.awt.Image.SCALE_SMOOTH);
+                final BufferedImage bufferedOutputImage = new BufferedImage(minimalWidth, minimalHeight, bufferedSourceImage.getType());
+                bufferedOutputImage.getGraphics().drawImage(upscaledImage, 0, 0, null);
+                final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+                ImageIO.write(bufferedOutputImage, formatName, outputStream);
```

Comment:
Can we please clarify which image format is used here? My concern is that any compressed format, such as JPEG, will use default compression, which may result in low quality of image. In case of PNG no such issue is present. Is it really a good idea to use the same image format as the source image format?

Reply:
That is a good question but I am afraid it is, at least input, out of our control. The image taken by ZenId should not be changed to support signatures. But output could be PNG, which is accepted by iProov, see https://docs.iproov.com/docs/Content/ImplementationGuide/api/api-endpoints.htm#POST_/claim/enrol/image